examples/advanced/quantum_volume.py

```diff
@@ -136,7 +153,86 @@ def sample_heavy_set(circuit: cirq.Circuit,
     return num_in_heavy_set / repetitions
 
 
-def main(num_qubits: int, depth: int, num_repetitions: int, seed: int):
+def compile_circuit(circuit: cirq.Circuit,
+                    *,
+                    device: cirq.google.xmon_device.XmonDevice,
+                    deconstruct: PointOptimizer = None,
+                    optimize: Callable[[cirq.Circuit], cirq.Circuit] = None
+                   ) -> Tuple[cirq.Circuit, Dict[cirq.ops.Qid, cirq.ops.Qid]]:
+    """Compile the given model circuit onto the given device. This follows a
+    similar compilation method as described in
+    https://arxiv.org/pdf/1811.12926.pdf Appendix A, although it does not
+    necessarily use QisKit to do so.
+
+    Args:
+        circuit: The model circuit to compile.
+        device: The device to compile onto.
+        deconstruct: An optional function to deconstruct the model circuit's
+            gates down to the targer devices gate set.
+        optimize: An optional function to optimize the circuit's gates
+            after routing.
+
+    Returns: A tuple where the first value is the compiled circuit and the
+        second value is the final mapping from the model circuit to the compiled
+        circuit. The latter is necessary in order to preserve the measurement
+        order.
+
+    """
+    compiled_circuit = circuit.copy()
+    # Step 1: Unrolling. Descend into each gate's hierarchical definition and
+    # rewrite it in terms of lower-level gates that are in the target device
+    # gate set.
+    if deconstruct:
+        deconstruct(compiled_circuit)
+    # Step 2: Swap Mapping (Routing). Ensure the gates can actually operate on
+    # the target qubits given our topology.
+    swap_network = ccr.route_circuit(compiled_circuit,
+                                     ccr.xmon_device_to_graph(device),
+                                     algo_name='greedy')
+    compiled_circuit = swap_network.circuit
+    # Step 3: Optimize. The paper uses various optimization techniques - because
+    # Quantum Volume is intended to test those as well, we allow this to be
+    # passed in. This optimizer is not allowed to change the order of the
+    # qubits.
+    if optimize:
+        compiled_circuit = optimize(compiled_circuit)
+
+    return (compiled_circuit, swap_network.final_mapping())
+
+
+class QuantumVolumeResult:
```

Comment:
I'd recommend "transposing" your classes.

What does this mean? Imagine your data as a table with each repetition being a row and each (model_circuit, heavy_set, ...) being a column. `QuantumVolumeResult` is a list (essentially) of columns. You'll find it more easy to wrangle data if instead you have a list of rows, where each row can be represented with a custom class

```python
@dataclass
class result:
  model_circuit
  heavy_set
  compiled_circuit
  sampler_result

def main():
  results = []
  for _ in range(reps):
    results.append(result(
      model_circuit=...
    ))
```

Bonus: now your (data)class can be immutable. 

Bonus: your invariant that all your current lists have to be the same length is trivially enforced

Reply:
Ah that's cleaner. I was thinking that in the end I wanted to print out this one large object, but having something that behaves as a bunch of rows will be nicer.